x/sequencer/keeper/rotation.go

```diff
@@ -0,0 +1,136 @@
+package keeper
+
+import (
+	"sort"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dymensionxyz/dymension/v3/x/sequencer/types"
+)
+
+func (k Keeper) startNoticePeriodForSequencer(ctx sdk.Context, seq *types.Sequencer) time.Time {
+	completionTime := ctx.BlockTime().Add(k.NoticePeriod(ctx))
+	seq.NoticePeriodTime = completionTime
+
+	k.UpdateSequencer(ctx, *seq)
+	k.AddSequencerToNoticePeriodQueue(ctx, *seq)
+
+	nextSeq := k.ExpectedNextProposer(ctx, seq.RollappId)
+	if nextSeq.SequencerAddress == "" {
+		k.Logger(ctx).Debug("rollapp will be left with no proposer after notice period", "rollappId", seq.RollappId, "sequencer", seq.SequencerAddress)
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeNoticePeriodStarted,
+			sdk.NewAttribute(types.AttributeKeyRollappId, seq.RollappId),
+			sdk.NewAttribute(types.AttributeKeySequencer, seq.SequencerAddress),
+			sdk.NewAttribute(types.AttributeKeyNextProposer, nextSeq.SequencerAddress),
+			sdk.NewAttribute(types.AttributeKeyCompletionTime, completionTime.String()),
+		),
+	)
+
+	return completionTime
+}
+
+// MatureSequencersWithNoticePeriod start rotation flow for all sequencers that have finished their notice period
+// The next proposer is set to the next bonded sequencer
+// The hub will expect a "last state update" from the sequencer to start unbonding
+func (k Keeper) MatureSequencersWithNoticePeriod(ctx sdk.Context, currTime time.Time) {
+	seqs := k.GetMatureNoticePeriodSequencers(ctx, currTime)
+	for _, seq := range seqs {
+		if k.IsProposer(ctx, seq.RollappId, seq.SequencerAddress) {
+			k.StartRotation(ctx, seq.RollappId)
+			k.removeNoticePeriodSequencer(ctx, seq)
+		}
+	}
+}
+
+// IsRotating returns true if the rollapp is currently rotating proposers
+func (k Keeper) IsRotating(ctx sdk.Context, rollappId string) bool {
+	return k.HasNextProposer(ctx, rollappId)
+}
+
+// IsNoticePeriodRequired returns true if the sequencer requires a notice period before unbonding
+// Both the proposer and the next proposer require a notice period
+func (k Keeper) IsNoticePeriodRequired(ctx sdk.Context, seq types.Sequencer) bool {
+	return k.IsProposer(ctx, seq.RollappId, seq.SequencerAddress) || k.IsNextProposer(ctx, seq.RollappId, seq.SequencerAddress)
+}
+
+// ExpectedNextProposer returns the next proposer for a rollapp
+func (k Keeper) ExpectedNextProposer(ctx sdk.Context, rollappId string) types.Sequencer {
+	// if nextProposer is set, were in the middle of rotation
+	seq, ok := k.GetNextProposer(ctx, rollappId)
+	if ok {
+		return seq
+	}
+
+	seqs := k.GetSequencersByRollappByStatus(ctx, rollappId, types.Bonded)
+	if len(seqs) == 0 {
+		return types.Sequencer{}
+	}
+
+	// take the next bonded sequencer to be the proposer. sorted by bond
+	sort.SliceStable(seqs, func(i, j int) bool {
```

Comment:
I think there is an edge case here where you may have a sequencer that is in "reduce bond" period, but it's still not reflected in his Tokens amount. 
In that case imo you want to already include the potential reduct in the highest bond calculation. 
more details here: https://www.notion.so/dymension/ADR-x-Sequencer-Bond-Management-b0a7061435fb40a28bd3cf64b899e968?pvs=4#8f0733cf6d2c427c87c07aeb522a7dff

Reply:
should be adjusted in the bond management PR.
out of scope for this PR